ddl/ddl_api.go

```diff
@@ -3251,6 +3252,30 @@ func (d *ddl) AddTablePartitions(ctx sessionctx.Context, ident ast.Ident, spec *
 		return errors.Trace(err)
 	}
 
+	if d.IsTiFlashPollEnabled() && meta.TiFlashReplica != nil {
+		// Must set placement rule before the ActionAddTablePartition job is in queue.
+		tikvStore, ok := ctx.GetStore().(helper.Storage)
+		if !ok {
+			logutil.BgLogger().Error("can not get Helper")
+		} else {
+			tikvHelper := &helper.Helper{
+				Store:       tikvStore,
+				RegionCache: tikvStore.GetRegionCache(),
+			}
+			for _, p := range partInfo.Definitions {
+				logutil.BgLogger().Info("add partition for TiFlash", zap.Int64("ID", p.ID))
+				ruleNew := MakeNewRule(p.ID, meta.TiFlashReplica.Count, meta.TiFlashReplica.LocationLabels)
+				if e := tikvHelper.SetPlacementRule(*ruleNew); e != nil {
+					return errors.Trace(err)
+				}
+				err := tikvHelper.PostAccelerateSchedule(p.ID)
+				if err != nil {
+					return errors.Trace(err)
+				}
+			}
+		}
+	}
+
```

Comment:
It is typically done in the another function, `onCreateTable`. The first step of DDL job is like validation and construct the job, while the actual action is done in the second step. I don't quite understand `Must set placement rule before the ActionAddTablePartition job is in queue.`, why is that?

Reply:
If there exists TiFlash replica for this table,  we must ensure the placement rule for TiFlash is successfully set before we actually do this DDL. If pd rule fails, we should abort the DDL.

Reply:
If your point is `If pd rule fails, we should abort the DDL.` Then you could put it in `onCreateTable`. Check this [PR](https://github.com/pingcap/tidb/issues/28159). It is enough to do something like:

```
onCreateTable(..)
...
 err = infosync.SetTiFlashReplica(context.TODO(), bundles) 
 if err != nil { 
 	job.State = model.JobStateCancelled  // this line abort the DDL job
 	return ver, errors.Wrapf(err, "failed to notify PD the placement rules") 
 }
```

Probably with some retrying to debounce network issues.